### ISSUE TITLE
Error Handling - LinkException 

### ISSUE THREAD

**Issue by pmcavoy89** [gql-dart/ferry#490]
So, I'm trying to do error handling with unauthenticated in my Flutter app. I'm getting back the correct response from my GraphQL server;
```
{"errors":[{"message":"User is not authenticated","extensions":{"code":"UNAUTHENTICATED","stacktrace":["GraphQLError: User is not authenticated"," ...
```

But when I go to try to access the response in my dart code with Ferry, I'm having issues with the following;
```dart
client
  .request(resourceReq)
  .listen((OperationResponse<GResourceData, GResourceVars> res) {
    if (res.hasErrors && res.linkException != null &&
          res.linkException is ServerException) {
        print("${res.linkException}");
    }
    ...
  }).onError((error) {
    print("Error $error");
  });
```

The code goes into the `onData` in the `.listen` and `res.hasError` is true but then `res.graphqlErrors` is null. I have figured out that the `hasError` checked the `res.linkException` which contains the following;
```
ServerException(originalException: null, originalStackTrace: null, parsedResponse: Response(data: null, errors: [GraphQLError(message: User is not authenticated, locations: null, path: null, extensions: {code: UNAUTHENTICATED, stacktrace: [GraphQLError: User is not authenticated ...
```
So, I can see it in the `ServerException` but I cannot access anything but `originalException` or `originalStackTrace` (which are null) as it is an `LinkException`. So, not sure how I can actually work with the `ServerException` that contains the `parsedResponse` yet when I try to do; `print("${res.linkException?.parsedResponse}");`, I get the following;

```
The getter 'parsedResponse' isn't defined for the type 'LinkException'.
Try importing the library that defines 'parsedResponse', correcting the name to the name of an existing getter, or defining a getter or field named 'parsedResponse'
```

Though that makes no sense as the if statement I have contains a type check to be sure it is of type `ServerException`, which does have `parsedResponse`. Any ideas?

**Comment by pmcavoy89** [comment]
Hm... seems like this solved it;

```dart
print("${(res.linkException as HttpLinkServerException).parsedResponse}");
```

Is there any documentation around this or a better way we can expose the error?

**Comment by knaeckeKami** [comment]
Does the server send an HTTP status code > 300 for this error?

the httpLink only parses requests with HTTP status code >= 200 <300, otherwise it throws an exception. 

See https://github.com/gql-dart/gql/blob/master/links/gql_http_link/lib/src/link.dart#L80

**Comment by pmcavoy89** [comment]
Yup. It throws a 401. That's what made me think I needed to case the `res.linkException` with `HttpLinkServerException`. 

```
throw new GraphQLError('User is not authenticated', {
  extensions: {
    code: 'Unauthenticated',
    http: { status: 401 },
  },
});
```

**Comment by knaeckeKami** [comment]
Ok.

It is recommended to not use HTTP status codes 4xx or 5xx in situations where you could use a well-formed graphql response, e.g something like 

```json
{
"data": null,
"errors": [ {"message": "User is not authenticated", "code": "..."}] 
}
```

There is no definitive standard for this, as graphql is not bound to a particular transport method (HTTP, websockets... ), but there is this spec: https://github.com/graphql/graphql-over-http

So for Content-Types of `application/json`, which the HttpClient uses, the spec says

> The server SHOULD NOT use a 4xx or 5xx status code.

This is why HttpLinks throws an Exception instead of returning a `Response()` object with the parsed errors, which is why your graphqlErrors list is null and the error details are only in your `LinkException`.

HttpLink should probably make this configurable though.

What you can do here instead:

- put an ErrorLink() before your HttpLink() so you can turn these Exceptions into Responses with proper graphlErrors
- just cast your exceptions to HttpLinkServerException, but then you are relying on the current implementation of your terminal link
- open a PR in the gql-dart repo to add configuration in the HttpLink on which status code are expected (which should return a normal Response if well-formed)  and which are unexpected errors which should always be treated a exception

**Comment by dunatron** [comment]
You shouldn't really be throwing errors for graphQL servers if you can help it IMO. 
Errors are schemaless and hard to keep track of. 

Consider using Unions instead, that way it makes it easier for an application to respond to what has happened. 
Your application is then not hijacked by REST status codes and you can make objects of types that make sense for your application

```
interface BaseError {
  message: String!
}

type UnknownError implements BaseError {
  message: String!
}

type NotAllowedError implements BaseError {
  message: String!
}

type UsersPage {
  count: Int!
  edges: [UserEdge!]!
  pageInfo: PageInfo!
}

union UsersPageResult =
    UsersPage
  | NotAuthed
  | NotAllowedError
  | UnknownError
  
type Query {
  users(input: UserCursorPaginationInput!): UsersPageResult!
}
```
